specs/accounting_reform/beacon-chain.md

```diff
@@ -0,0 +1,278 @@
+# Incentive accounting  reforms
+
+## Table of contents
+
+**TODO**: Generate table of contents with doctoc.
+
+## Introduction
+
+This is a proposed incentive accounting simplification which replaces `PendingAttestation` with epoch participation flags.
+
+## Constants
+
+### Participation flags
+
+| Name | Value |
+| - | - |
+| `TIMELY_HEAD_FLAG` | `0` |
+| `TIMELY_SOURCE_FLAG` | `1` |
+| `TIMELY_TARGET_FLAG` | `2` |
+
+### Participation rewards
+
+| Name | Value |
+| - | - |
+| `TIMELY_HEAD_NUMERATOR` | `12` |
+| `TIMELY_SOURCE_NUMERATOR` | `12` |
+| `TIMELY_TARGET_NUMERATOR` | `32` |
+| `REWARD_DENOMINATOR` | `64` |
+
+The reward fractions add up to 7/8, leaving the remaining 1/8 for proposer rewards and other future micro-rewards.
+
+### Misc
+
+| Name | Value |
+| - | - |
+| `PARTICIPATION_FLAGS_LENGTH` | `8` |
+| `FLAGS_AND_NUMERATORS` | `((TIMELY_HEAD_FLAG, TIMELY_HEAD_NUMERATOR), (TIMELY_SOURCE_FLAG, TIMELY_SOURCE_NUMERATOR), (TIMELY_TARGET_FLAG, TIMELY_TARGET_NUMERATOR))` |
+
+## Containers
+
+### Modified containers
+
+#### `BeaconState`
+
+```python
+class BeaconState(Container):
+    # Versioning
+    genesis_time: uint64
+    genesis_validators_root: Root
+    slot: Slot
+    fork: Fork
+    # History
+    latest_block_header: BeaconBlockHeader
+    block_roots: Vector[Root, SLOTS_PER_HISTORICAL_ROOT]
+    state_roots: Vector[Root, SLOTS_PER_HISTORICAL_ROOT]
+    historical_roots: List[Root, HISTORICAL_ROOTS_LIMIT]
+    # Eth1
+    eth1_data: Eth1Data
+    eth1_data_votes: List[Eth1Data, EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH]
+    eth1_deposit_index: uint64
+    # Registry
+    validators: List[Validator, VALIDATOR_REGISTRY_LIMIT]
+    balances: List[Gwei, VALIDATOR_REGISTRY_LIMIT]
+    # Randomness
+    randao_mixes: Vector[Bytes32, EPOCHS_PER_HISTORICAL_VECTOR]
+    # Slashings
+    slashings: Vector[Gwei, EPOCHS_PER_SLASHINGS_VECTOR]  # Per-epoch sums of slashed effective balances
+    # Participation
+    previous_epoch_participation: List[Bitvector[PARTICIPATION_FLAGS_LENGTH], VALIDATOR_REGISTRY_LIMIT]
```

Comment:
I feel like we should swap the order of the validator and flag indices here (make it `List[Bitlist[VALIDATOR_REGISTRY_LIMIT], PARTICIPATION_FLAGS_LENGTH]`): Currently Bitvectors are padded to full bytes when serializing or Merkleizing, meaning that each 3 bit vector will take up a full byte of space.

Reply:
The problem I see with doing that it that it would increase the amount of Merkling you need to do to process an attestation by ~3x, as you would have to adjust the branches in three positions instead of one. I think on net it would be a reduction in efficiency (as a Merkle branch going into a committee is in most cases longer than the committee itself).

Reply:
I see, We could also do the following:
1. Change the SSZ spec to allow for tight packing of Bitlists (Don't think we have used vectors of Bitlists, so this should be fine)
2. Make it a Bitlist[PARTICIPATION_FLAGS_LENGTH * VALIDATOR_REGISTRY_LIMIT] and address manually.

I think 2 would be easy to do and would suggest that.